src/core/Instrument.cpp

```diff
@@ -78,8 +82,83 @@ bool Instrument::isFromTrack( const Track * _track ) const
 	return( m_instrumentTrack == _track );
 }
 
+int count_zero_crossings(sampleFrame *buf,fpp_t start, fpp_t frames)
+{
+	// zero point crossing counts of all channels
+	int zero_crossings[DEFAULT_CHANNELS] = {0};
+	// maximum zero point crossing of all channels
+	int max_zc = 0;
+
+	// determine the zero point crossing counts
+	for (fpp_t f = start; f < frames; ++f)
+	{
+		for (ch_cnt_t ch=0; ch < DEFAULT_CHANNELS; ++ch)
+		{
+			// we don't want to count [-1, 0, 1] as two crossings
+			if ((buf[f-1][ch] <= 0.0 && buf[f][ch] > 0.0) ||
+					(buf[f-1][ch] >= 0.0 && buf[f][ch] < 0.0))
+			{
+				++zero_crossings[ch];
+				if (zero_crossings[ch] > max_zc)
+				{
+					max_zc = zero_crossings[ch];
+				}
+			}
+		}
+	}
+
+	return max_zc;
+}
+
+
+void Instrument::applyFadeIn(sampleFrame * buf, NotePlayHandle * n)
+{
+	f_cnt_t total = n->totalFramesPlayed();
+	if (total == 0)
+	{
+		const fpp_t frames = n->framesLeftForCurrentPeriod();
+
+		int max_zc = count_zero_crossings(buf, 1, frames);
+
+		// calculate the length of the fade in
+		// Length is inversely proportional to the max of zero_crossings,
+		// because for low frequencies, we need a longer fade in to
+		// prevent clicking.
+		fpp_t length = (fpp_t) (85.0  / ((float)max_zc / ((float)frames/128.0f) + 1.0f));
+		n->m_fadeInLength = length;
 
+		// apply fade in
+		length = length < frames ? length : frames;
+		for (fpp_t f = 0; f < length; ++f)
+		{
+			for (ch_cnt_t ch = 0; ch < DEFAULT_CHANNELS; ++ch)
+			{
+				buf[f][ch] *= 0.5 - 0.5 * cosf(F_PI * (float) f / (float)n->m_fadeInLength);
+			}
+		}
+	}
+	else if(total < n->m_fadeInLength)
+	{
+		const fpp_t frames = n->framesLeftForCurrentPeriod();
+
+		int new_zc = count_zero_crossings(buf, 0, frames);
+
+		if(new_zc != 0)
+		{
+			n->m_fadeInLength = (fpp_t) ((float)n->m_fadeInLength * 0.875f);
```

Comment:
So, I guess `7/8` is just a guess for what sounded good? You're just saying: If the wave crosses 0 again, we can increase speed a bit?

It might work in this direction, but imagine the opposite: The note starts with tons of zero crossings, so the algorithm thinks "I can make a quick fade in". In the second buffer it then sees "not many/no zero crossings", so it actually increased too fast. In this case, shouldn't you slow down?

Though I just think there are cases where you have no chance. Imagine you have 32 samples buffersize, and the algorithm calculates a fade of 20 for the first 32 samples. It's finished then. In the next 32 sample, the number of zero crossings is way lower. Did the algorithm increase too fast?

Maybe you could rather do it like this: Calculate the expected fade in from 0 to the last sample (i.e. `total + length`) in each call of `applyFadeIn`. Calculate the increase speed for this buffer depending on that. Interpolate the new speed respective to the old to not have a steep edge. But I'm not sure at all if this is better.

Reply:
> ... In this case, shouldn't you slow down?

Since the purpose of this algorithm is to prevent clicking while producing punchy notes, I think it shouldn't. If a signal starts with lots of zero crossings followed by very few zero crossings, it means that it has a harsh beginning. We don't want to over-smooth the wave, that's the purpose of the regular envelope. We just want to prevent clicks in this context.

At 32 samples, even if the algorithms sees only one zero point crossing, it generates a fade in of length 17 samples. At 64 samples, it fades-in in 28.33 samples for one zero point crossing. So it's always too late to slow down.

> Interpolate the new speed respective to the old to not have a steep edge

If we interpolate using `currentSpeed = (oldSpeed+newSpeed)/2.0`, It might be harsher than the current one. Perhaps something like `currentSpeed = (3*oldSpeed+newSpeed)/4.0` can work better. Interpolation is a great idea, I need to experiment.

Reply:
You could also try something like `currentSpeed = f * newSpeed + (blockSize-f) * oldSpeed` (where `f` is the frame running from `0` to `blockSize - 1`).